feat: add repositories to workspace in bulk

zeucapua opened the pull request:
## Description

Add a new option to `TrackedReposWizard` to allow users to add repositories by pasting them into a text area, each separated into new lines. User can then filter from the added repositories before finalizing the import.

NOTE: repositories are case sensitive when trying to fetch during `Update Workspace`. I added a note on the modal above where the text are so users can see before pasting their list. This means that if there is even only 'one' repo with the wrong cAse, it'll error out adding the entire list. How should we address this? @open-sauced/engineering 

<!-- 
Please do not leave this blank 
This PR [adds/removes/fixes/replaces] the [feature/bug/etc]. 
-->

## What type of PR is this? (check all applicable)

- [X] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Closes #2603 
<!-- 
Please use this format link issue numbers: Fixes #123
https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 
-->

## Mobile & Desktop Screenshots/Recordings

https://github.com/open-sauced/app/assets/20603494/0bd8f8f9-6056-48cf-81dd-a0faba11d2b1

<!-- Visual changes require screenshots -->


## Steps to QA

1. Go to `/workspaces/new` or `/workspaces/settings`
2. Click `Add repositories`
3. Paste a list of repositories (CASE SENSITIVE)
4. Filter through the pasted repositories
5. Finish modal and click `Update Workspace`

<!-- 
Please provide some steps for the reviewer to test your change. If you have wrote tests, you can mention that here instead.

1. Click a link
7. Do this thing
8. Validate you see the thing working
-->


## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?



## [optional] What gif best describes this PR or how it makes you feel?

![list-gif](https://github.com/open-sauced/app/assets/20603494/01c2b1e6-2d53-462c-9015-7aa3214998af)


<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/.github/blob/main/CONTRIBUTING.md.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

nickytonline commented:
Let us know if this icon is OK @isabensusan 

![CleanShot 2024-02-06 at 19 50 53](https://github.com/open-sauced/app/assets/833231/8773b34b-b97d-4e6d-b0d5-7ffa94e26377)

bdougie commented:
Reading the note about the cAse.

I am curious if there is a need for case sensitivity and if we can lowercase (sanitize) it all to remove the burden from the user.

jpmcb commented:
From the APIs perspective, we do sanitize case sensitivity, so when adding repos, it shouldn't be necessary to have the users make repos case sensitive. And from GitHub's perspective, repo names (and most URL parameters, like usernames) are also not case sensitive. Example: https://github.com/Open-Sauced/APP is the same as https://github.com/open-sauced/app

jpmcb commented:
If there's an error when attempting to fetch repos during update workspace with case sensitive repos, that sounds like a bug 🐛

zeucapua commented:
Went to the preview deploy, now I'm having different behavior. It's still not adding not correctly cased repositories, but it's adding those that are. 

See video, only `TanStack/router` is correctly cased. Clearly the table gets all the repos, just on `Update Workspace` something is going wrong. I can look into it in the morning.

https://github.com/open-sauced/app/assets/20603494/343300a4-2740-4447-913f-f9fb4b5c7e47

nickytonline commented on components/Workspaces/TrackedRepoWizard/PasteReposStep.tsx:
```diff
@@ -0,0 +1,31 @@
+export const PasteReposStep = ({ onBulkAddRepos }: PasteReposStepProps) => {
+  const [pastedInput, setPastedInput] = useState("");
+
+  const parseInput = () => {
```
refactor (blocking): if the textarea is empty return instead of doing the bulk import. Otherwise you get to the next step with a repository that is an empty string.

![CleanShot 2024-02-07 at 09 51 22](https://github.com/open-sauced/app/assets/833231/30d11706-6f70-4d9a-919e-92105ce4a654)

Until the value of the textarea has text, the import button should be disabled.

nickytonline commented on components/Workspaces/TrackedRepoWizard/PasteReposStep.tsx:
```diff
@@ -0,0 +1,31 @@
+        value={pastedInput}
+        onChange={(e) => setPastedInput(e.target.value)}
+        placeholder="Paste repositories here line by line (e.g. open-sauced/app)"
+        className="p-4 border rounded-xl h-full"
```
When the textarea gets focus, it uses the default blue outline of the browser. There are custom focus styles already set in the application. It's a little more involved to add it to the global styles for now, so in the interim, you can add the focus styles to the textarea in this component.

```
outline-none focus-visible:ring focus-visible:border-orange-500 focus-visible:ring-orange-100
```

nickytonline reviewed (changes requested):
This is working great @zeucapua! A couple change requests.

nickytonline commented:
I've ignored the casing issue mentioned as it looks like @jpmcb is seeing if it's an issue in the API.

nickytonline commented:
> I opened a PR in the API with a fix for normalizing the case when searching for repos.
> 
> 
> 
> https://github.com/open-sauced/api/pull/547

Cool cool. So we can remove the red text warning then @zeucapua

zeucapua commented:
Added validation and sanitation checks to:
- disable importing if the textarea is empty (even with empty lines)
- remove any empty lines between the repos that can result in an empty fetch

Also changed styling so the case sensitivity note is gone and the textarea on focus is more on brand. Attached a video to see it all in action:

https://github.com/open-sauced/app/assets/20603494/fb5a709f-ee2c-4317-a54e-878ff0633aea

nickytonline commented on components/Workspaces/TrackedRepoWizard/FilterPastedReposStep.tsx:
```diff
@@ -0,0 +1,59 @@
+
+  return (
+    <div className="flex flex-col gap-4 h-96 max-h-96">
+      {repositories.size === 0 ? (
+        <EmptyState />
+      ) : (
```
I don't think the empty state us required as we'll always be coming from the pasted textarea that has at least one repo in it.

nickytonline reviewed (approved):
Works great @zeucapua! thanks for making the changes!

🚀

brandonroberts reviewed (approved):
🍕
